mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp

```diff
@@ -82,22 +90,43 @@ func::FuncOp AdjointLowering::discardAndReturnReg(PatternRewriter &rewriter, Loc
         rewriter.setInsertionPointAfter(callee);
         unallocFn =
             rewriter.create<func::FuncOp>(loc, fnName, fnType, visibility, nullptr, nullptr);
-        // clone the body.
+        // Clone the body.
+        // TODO: we do not support submodules, i.e. our gradient functions
+        // will have just one block.
+        assert(callee.getBody().hasOneBlock() &&
+               "Gradients with quantum submodules are not supported");
         rewriter.cloneRegionBefore(callee.getBody(), unallocFn.getBody(), unallocFn.end());
         rewriter.setInsertionPointToStart(&unallocFn.getBody().front());
 
-        // Let's capture the qreg.
-        quantum::DeallocOp localDealloc = *unallocFn.getOps<quantum::DeallocOp>().begin();
-
         // Let's return the qreg and erase the device release.
-        unallocFn.walk([&](Operation *op) {
-            if (isa<quantum::DeviceReleaseOp>(op)) {
-                rewriter.eraseOp(op);
-            }
-            else if (isa<func::ReturnOp>(op)) {
-                op->setOperands(localDealloc.getOperand());
-            }
-        });
+        SmallVector<Value> returnVals;
+
+        // Fine for now: only one block in body so only one quantum.dealloc
+        auto localDealloc = *unallocFn.getOps<quantum::DeallocOp>().begin();
```

Comment:
Grab the terminator directly, it's faster faster than iterating to find ops of a particular type :)